app/controllers/two_factor_authentication/sms_opt_in_controller.rb

```diff
@@ -0,0 +1,97 @@
+module TwoFactorAuthentication
+  class SmsOptInController < ApplicationController
+    before_action :load_phone_configuration
+
+    def new
+      @other_mfa_options_url = other_options_mfa_url
+      @cancel_url = cancel_url
+
+      analytics.track_event(
+        Analytics::SMS_OPT_IN_VISIT,
+        new_user: new_user?,
+        has_other_auth_methods: has_other_auth_methods?,
+        phone_configuration_id: @phone_configuration.id,
+      )
+    end
+
+    def create
+      response = opt_out_manager.opt_in_phone_number(@phone_configuration.formatted_phone)
+
+      analytics.track_event(
+        Analytics::SMS_OPT_IN_SUBMITTED,
+        response.to_h.merge(
+          new_user: new_user?,
+          has_other_auth_methods: has_other_auth_methods?,
+          phone_configuration_id: @phone_configuration.id,
+        ),
+      )
+
+      if response.success?
+        redirect_to otp_send_url(otp_delivery_selection_form: { otp_delivery_preference: :sms })
+      else
+        @other_mfa_options_url = other_options_mfa_url
+        @cancel_url = cancel_url
+
+        if !response.error
+          # unsuccessful, but didn't throw an exception: already opted in last 30 days
+          render :error
+        else
+          # one-off error, show form so users can try again
+          flash[:error] = t('two_factor_authentication.opt_in.error_retry')
+          render :new
+        end
+      end
+    end
+
+    private
+
+    def opt_out_manager
+      @opt_out_manager ||= Telephony::Pinpoint::OptOutManager.new
```

Comment:
What benefit is the memoization here?

Reply:
I guess since we use it once, none, it was just habit. Will remove
```suggestion
      Telephony::Pinpoint::OptOutManager.new
```